### PR TITLE
[Dynamo] Log backward graph compilation metrics

### LINKED ISSUE #125313: Dynamo doesn't log backward graphs in compilation metrics
### 🐛 Describe the bug

Repro:
```
import torch
import torch.nn.functional as F

class BasicModule(torch.nn.Module):
    def __init__(self):
        super().__init__()
        self.linear1 = torch.nn.Linear(10, 10)
        self.scale = torch.randn(1, 10)

    def forward(self, x):
        x = self.linear1(x)
        torch._dynamo.graph_break()
        return F.relu(x) * self.scale

x = torch.ones(1, 10)
mod = BasicModule()
mod = torch.compile(mod)
print(mod(x))

mod(x).sum().backward()
```

with the following changes:
```
diff --git a/torch/_inductor/compile_fx.py b/torch/_inductor/compile_fx.py
index 7f3b2ec95f..b98dabb6f0 100644
--- a/torch/_inductor/compile_fx.py
+++ b/torch/_inductor/compile_fx.py
@@ -420,6 +420,7 @@ def compile_fx_inner(
     layout_opt: Optional[bool] = None,
     extern_node_serializer: Optional[Callable[[List[ExternKernelNode]], Any]] = None,
 ) -> Union[CompiledFxGraph, str]:
+    print("running compile_fx_inner")
     """
     Inductor API that compiles a single graph.
 
diff --git a/torch/_utils_internal.py b/torch/_utils_internal.py
index 3da8bc2186..7dcd65e4f5 100644
--- a/torch/_utils_internal.py
+++ b/torch/_utils_internal.py
@@ -92,7 +92,7 @@ def signpost_event(category: str, name: str, parameters: Dict[str, Any]):
 
 
 def log_compilation_event(metrics):
-    log.info("%s", metrics)
+    print("%s", metrics)
```

Output:
```
running compile_fx_inner
%s CompilationMetrics(frame_key='1', co_name='forward', co_filename='/data/users/ybliang/debug/debug2.py', co_firstlineno=10, cache_size=0, accumulated_cache_size=0, guard_count=11, shape_env_guard_count=0, graph_op_count=1, graph_node_count=3, graph_input_count=1, start_time=1716076212.7113512, entire_frame_compile_time_s=6.625730991363525, backend_compile_time_s=6.587674379348755, inductor_compile_time_s=4.015706300735474, code_gen_time_s=3.9247825145721436, fail_type=None, fail_reason=None, fail_user_frame_filename=None, fail_user_frame_lineno=None, non_compliant_ops=set(), compliant_custom_ops=set(), restart_reasons={"'skip function graph_break in file /home/ybliang/local/pytorch/torch/_dynamo/decorators.py'"}, dynamo_time_before_restart_s=0.0244600772857666, has_guarded_code=True)
running compile_fx_inner
%s CompilationMetrics(frame_key='2', co_name='torch_dynamo_resume_in_forward_at_12', co_filename='/data/users/ybliang/debug/debug2.py', co_firstlineno=12, cache_size=0, accumulated_cache_size=0, guard_count=10, shape_env_guard_count=0, graph_op_count=2, graph_node_count=5, graph_input_count=1, start_time=1716076219.3381996, entire_frame_compile_time_s=0.10369133949279785, backend_compile_time_s=0.09059739112854004, inductor_compile_time_s=0.03840017318725586, code_gen_time_s=0.02279829978942871, fail_type=None, fail_reason=None, fail_user_frame_filename=None, fail_user_frame_lineno=None, non_compliant_ops=set(), compliant_custom_ops=set(), restart_reasons=set(), dynamo_time_before_restart_s=0.0, has_guarded_code=True)
tensor([[-0.5529,  0.0000, -0.1201, -0.0000, -0.0802,  0.1059, -0.1784, -0.0000,
         -0.0401,  0.0000]], grad_fn=<CompiledFunctionBackward>)
running compile_fx_inner
running compile_fx_inner
```

```compile_fx_inner``` has been called twice, forward and backward respectively. However, we log the compilation metrics only for forward graph. This is because the backward graph Inductor compilation is not captured by ```convert_frame```. This causes confusion when debugging compilation time related issues, since the metrics are not accurate. 

### Versions

N/A

cc @ezyang @msaroufim @bdhirsh @anijain2305 @chauhang @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @chenyang78 @kadeng

### PR DESCRIPTION
**Pull request by yanboliang**
Fixes #125313

Compilation metric logs for the code example at #125313:
```
%s CompilationMetrics(compile_id='0/0', frame_key='1', co_name='forward', co_filename='/data/users/ybliang/debug/debug2.py', co_firstlineno=10, cache_size=0, accumulated_cache_size=0, guard_count=11, shape_env_guard_count=0, graph_op_count=1, graph_node_count=3, graph_input_count=1, start_time=1716247236.6165977, entire_frame_compile_time_s=7.926939964294434, backend_compile_time_s=7.887059926986694, inductor_compile_time_s=4.108498811721802, code_gen_time_s=3.97833514213562, fail_type=None, fail_reason=None, fail_user_frame_filename=None, fail_user_frame_lineno=None, non_compliant_ops=set(), compliant_custom_ops=set(), restart_reasons={"'skip function graph_break in file /home/ybliang/local/pytorch/torch/_dynamo/decorators.py'"}, dynamo_time_before_restart_s=0.025330543518066406, has_guarded_code=True, is_fwd=True)
%s CompilationMetrics(compile_id='1/0', frame_key='2', co_name='torch_dynamo_resume_in_forward_at_12', co_filename='/data/users/ybliang/debug/debug2.py', co_firstlineno=12, cache_size=0, accumulated_cache_size=0, guard_count=10, shape_env_guard_count=0, graph_op_count=2, graph_node_count=5, graph_input_count=1, start_time=1716247244.544928, entire_frame_compile_time_s=0.10148310661315918, backend_compile_time_s=0.08753013610839844, inductor_compile_time_s=0.03691983222961426, code_gen_time_s=0.022417306900024414, fail_type=None, fail_reason=None, fail_user_frame_filename=None, fail_user_frame_lineno=None, non_compliant_ops=set(), compliant_custom_ops=set(), restart_reasons=set(), dynamo_time_before_restart_s=0.0, has_guarded_code=True, is_fwd=True)
tensor([[-0.1622, -0.0000, -0.0000,  0.5643, -0.0000,  0.0000, -0.5087,  0.0914,
         -0.0000, -0.0421]], grad_fn=<CompiledFunctionBackward>)
%s CompilationMetrics(compile_id='1/0', frame_key=None, co_name=None, co_filename=None, co_firstlineno=None, cache_size=None, accumulated_cache_size=None, guard_count=None, shape_env_guard_count=None, graph_op_count=None, graph_node_count=None, graph_input_count=None, start_time=None, entire_frame_compile_time_s=None, backend_compile_time_s=None, inductor_compile_time_s=0.026738643646240234, code_gen_time_s=0.016446352005004883, fail_type=None, fail_reason=None, fail_user_frame_filename=None, fail_user_frame_lineno=None, non_compliant_ops=None, compliant_custom_ops=None, restart_reasons=None, dynamo_time_before_restart_s=None, has_guarded_code=None, is_fwd=False)
%s CompilationMetrics(compile_id='0/0', frame_key=None, co_name=None, co_filename=None, co_firstlineno=None, cache_size=None, accumulated_cache_size=None, guard_count=None, shape_env_guard_count=None, graph_op_count=None, graph_node_count=None, graph_input_count=None, start_time=None, entire_frame_compile_time_s=None, backend_compile_time_s=None, inductor_compile_time_s=0.14563536643981934, code_gen_time_s=0.08652091026306152, fail_type=None, fail_reason=None, fail_user_frame_filename=None, fail_user_frame_lineno=None, non_compliant_ops=None, compliant_custom_ops=None, restart_reasons=None, dynamo_time_before_restart_s=None, has_guarded_code=None, is_fwd=False)
```

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @peterbell10 @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @ColinPeppler @amjames @desertfire @chauhang

### CODE REVIEW

File: torch/_dynamo/utils.py [hunk 1]
```diff
@@ -620,19 +660,32 @@ class CompilationMetrics:
...
-_compilation_metrics: Deque[CompilationMetrics] = collections.deque(
-    maxlen=DEFAULT_COMPILATION_METRICS_LIMIT
-)
+_compilation_metrics: Deque[
+    Union[FwdCompilationMetrics, BwdCompilationMetrics]
+] = collections.deque(maxlen=DEFAULT_COMPILATION_METRICS_LIMIT)
 
 
-def record_compilation_metrics(compilation_metrics: CompilationMetrics):
+def record_compilation_metrics(
```
Comment (yanboliang) [inline torch/_dynamo/utils.py:678]: I'll do the fbcode changes to log bwd graph compilation metrics as a new row in Scuba table, but let's make the OSS version in a good shape firstly.

File: torch/_dynamo/utils.py [hunk 2]
```diff
@@ -177,6 +177,15 @@ def print_time_report():
     print(out)
 
 
+def _add_time_spent(key, phase_name, time_spent):
+    if key not in frame_phase_timing:
+        frame_phase_timing[key] = {}
```
Comment (Skylion007) [inline torch/_dynamo/utils.py:182]:
```suggestion
    frame_phase_timing.setdefault(key, {})
```

File: torch/_dynamo/utils.py [hunk 3]
```diff
@@ -195,7 +204,7 @@ def print_time_report():
 # The frame is incremented outside of this function, in def increment_frame() above.
 
 
-def dynamo_timed(original_function=None, phase_name=None):
+def dynamo_timed(original_function=None, phase_name=None, fwd_only=True):
```
Comment (ezyang) [inline torch/_dynamo/utils.py:207]: I couldn't tell from the use sites what this kwarg does, some docs would be appreciated

File: torch/_dynamo/utils.py [hunk 4]
```diff
@@ -620,19 +660,32 @@ class CompilationMetrics:
...
+):
     global _compilation_metrics
     _compilation_metrics.append(compilation_metrics)
+    if isinstance(compilation_metrics, FwdCompilationMetrics):
+        name = "fwd_compilation_metrics"
+    else:
+        name = "bwd_compilation_metrics"
     torch._logging.trace_structured(
-        "compilation_metrics",
+        name,
```
Comment (ezyang) [inline torch/_dynamo/utils.py:688]: I'm meh on this. In particular, tlparse currently knows how to parse `compilation_metrics` but not `fwd_compilation_metrics`. If you rename it this way, you have to adjust tlparse to know how to deal with both namings. But why not just not rename it.

File: torch/_dynamo/utils.py [hunk 5]
```diff
@@ -212,12 +221,42 @@ def time_wrapper(*args, **kwargs):
...
+                                ].get("inductor_compile", None)
+                                code_gen_time = frame_phase_timing[compile_id].get(
+                                    "code_gen", None
+                                )
+                                metrics = BwdCompilationMetrics(
+                                    compile_id,
+                                    inductor_compile_time,
+                                    code_gen_time,
+                                )
+                                record_compilation_metrics(metrics)
```
Comment (ezyang) [inline torch/_dynamo/utils.py:258]: This is copied from the code in AOTAutograd. But we shouldn't have two copies of this. Can you also delete the AOTAutograd code in this PR?

File: torch/_dynamo/utils.py (hunk 4 above)
Comment (ezyang) [reply torch/_dynamo/utils.py:688]: Ditto for the backwards naming. Like, you don't *have* to keep the names the same (tlparse is smart enough to ignore sections that it doesn't understand), but if you do rename it, you need to update tlparse

File: torch/_dynamo/utils.py (hunk 3 above)
Comment (yanboliang) [reply torch/_dynamo/utils.py:207]: Will add docs!

File: torch/_dynamo/utils.py (hunk 5 above)
Comment (yanboliang) [reply torch/_dynamo/utils.py:258]: No, this is not copied from AOT, can you advice where is the similar code in AOT? I can do refactor if there is.

File: torch/_dynamo/utils.py (hunk 4 above)
Comment (yanboliang) [reply torch/_dynamo/utils.py:688]: The main reason of having two separate metrics definition is there are too many Nones at backward metrics if we use the original ```CompilationMetrics```. But right now, I was convinced to not rename it and keep both fwd and bwd use the same definition class.

File: torch/_dynamo/utils.py (hunk 5 above)
Comment (ezyang) [reply torch/_dynamo/utils.py:261]: in torch/_functorch/_aot_autograd/jit_compile_runtime_wrappers.py search for `aot_autograd_backward_compilation_metrics`

File: torch/_dynamo/utils.py (hunk 4 above)
Comment (ezyang) [reply torch/_dynamo/utils.py]: Well hang on. What does the name here affect whether or not you have two separate dataclasses. I also think two separate dataclasses is appropriate lol

File: test/dynamo/test_structured_trace.py [hunk 6]
```diff
@@ -142,7 +142,7 @@ def test_schedule(self):
 {"inductor_output_code": {"filename": "FILENAME"}, "frame_id": 0, "frame_compile_id": 0, "attempt": 0, "has_payload": "HASH"}
 {"dynamo_guards": {}, "frame_id": 0, "frame_compile_id": 0, "attempt": 0, "has_payload": "HASH"}
 {"dynamo_cpp_guards_str": {}, "frame_id": 0, "frame_compile_id": 0, "attempt": 0, "has_payload": "HASH"}
-{"compilation_metrics": "METRICS", "frame_id": 0, "frame_compile_id": 0, "attempt": 0}
+{"fwd_compilation_metrics": "METRICS", "frame_id": 0, "frame_compile_id": 0, "attempt": 0}
```
Comment (ezyang) [inline test/dynamo/test_structured_trace.py:145]: You're still egregiously breaking BC for no good reason

**Review by ezyang — CHANGES_REQUESTED** [review]
You haven't told me WHY it's a good idea to rename compilation_metrics to fwd_compilation_metrics

File: test/dynamo/test_structured_trace.py (hunk 6 above)
Comment (yanboliang) [reply test/dynamo/test_structured_trace.py:145]: Because we collect these metrics only for fwd graphs, ```BwdCompilationMetrics``` is used to store compilation metrics for bwd graphs. I'm ok to keep ```CompilationMetrics``` and add ```BwdCompilationMetrics```, but not sure if this would confuse others who look at these metrics.

Comment (ezyang) [reply test/dynamo/test_structured_trace.py:145]: You're welcome to rename the struct, but don't rename the JSON key name, that's BC surface

Comment (yanboliang) [reply test/dynamo/test_structured_trace.py]: I'll keep the original ```CompilationMetrics``` for fwd and add ```BwdCompilationMetrics``` for bwd.

File: torch/_dynamo/utils.py [hunk 7]
```diff
@@ -185,6 +185,14 @@ def print_time_report():
     print(out)
 
 
+def _add_time_spent(key, phase_name, time_spent):
+    frame_phase_timing.setdefault(key, {})
+    if phase_name not in frame_phase_timing[key]:
+        frame_phase_timing[key][phase_name] = time_spent
+    else:
+        frame_phase_timing[key][phase_name] += time_spent
```
Comment (ezyang) [inline torch/_dynamo/utils.py:193]: nit: a defaultdict(int) would eliminate the need for the in graph test

**Review by ezyang — APPROVED** [review]
We can try this ig but I'm pretty suspicious of putting this logic in `dynamo_timed`. It seems far too easy to accidentally generate way more compilation metrics than you actually intended to generate. Do you have some way you're going to validate this change is actually generating the metrics you need?

**Comment by yanboliang** [comment]
> Do you have some way you're going to validate this change is actually generating the metrics you need?

@ezyang Actually the backward compilation metrics only being record ```if phase_name is not None```, which means it won't record these metrics if you just use ```dynamo_timed```. ```phase_name``` is added for logging metrics purpose, so I think this is a proper check that we should have.

**Comment by ezyang** [comment]
@yanboliang What I find confusing about the current setup is I wouldn't expect `@dynamo_timed` to be what is responsible for reporting overall metrics to Scuba

**Comment by yanboliang** [comment]
@ezyang I got your point, so probably I can have a separate decorator for recording metrics. We just need to split the logics out of the current ```dynamo_timed``` when ```phase_name is not None```. I'll send a PR for this if you like this idea.
